### PR TITLE
PubSub (UI + Backend)

### PR DESCRIPTION
**Pull request by yawri**
## Description

This PR integrates the frontend “PubSub Notifications” UI with the backend PubSub endpoints. Key changes include:

- **Frontend (`script.js`)**  
  - Replaced all `localStorage` lookups with real API calls to `/api/project/:id`, `/api/pubsub/update_newsletter_papers`, and `/api/pubsub/get_newsletter_papers`.  
  - On page load, fetch project metadata and render the header.  
  - After clearing the PubSub cards and wiring up the subscribe form, POST to `update_newsletter_papers` (tolerant of “no-queries” or empty results) to trigger backend candidate selection.  
  - GET the fresh list from `get_newsletter_papers?projectId=…`, log and render either real recommendations or fallback test cards.  
  - Removed stale PDF-upload-related and localStorage code comments that were no longer needed.

- **Backend (`app.py`)**  
  - Enhanced `/api/projects` to accept and store a `queries` array in the `projects_table`.  
  - Updated `/api/pubsub/update_newsletter_papers` to:
    - Early-exit with `status: no-queries` if the project has no saved queries.  
    - Catch Chroma-specific `ValueError` and return `status: no-results`.  
  - Made `/api/pubsub/get_newsletter_papers` accept both `projectId` and `project_id` query params for robustness.  

- **Database Handlers**  
  - Extended `add_new_project_to_db` to save the new `queries` column.  
  - Renamed and refactored the newsletter-tag reset function to accept `project_id` as an argument, so only that project’s flags are cleared.  
  - Verified that `get_pubsub_papers_for_project` returns only `newsletter = TRUE AND seen = FALSE` rows.

It connects the UI’s PubSub cards to the backend logic that selects, tags, and stores newsletter papers, replacing mock/test data with live recommendations.

Fixes #\<issue-number\>  

## Type of change

- [x] Refactor (code cleanup, remove commented sections, improve integration)
- [x] New feature (full end-to-end PubSub UI ↔ backend flow)

## How Has This Been Tested?

1. **Create & view project**  
   - Created a new project via `/create-project`, verified redirect to `/project/:id`.  
2. **PubSub flow**  
   - Confirmed `/api/pubsub/update_newsletter_papers` returns `status: ok`, and papers appear on `/api/pubsub/get_newsletter_papers`.  
   - Loaded the project page, saw the live PubSub cards instead of test placeholders.  
3. **Edge cases**  
   - Projects with no queries now return `no-queries` without errors.  
   - Projects with no new candidates return `no-results` and still fallback.  
4. **General UI & PDF upload**  
   - Uploaded a PDF to verify `extract-pdf-text` remains unaffected.  
   - Ran through the entire recommendations streaming flow to ensure no regressions.

## Reviewers

@AndresDChB  @andresCh01

### CODE REVIEW

**Comment by JonasBaeumer** [comment]
Lets not have the functionality where we wipe the database after each docker compose up because if the projects are not persistent what's the purpose of having multiple projects anyways.

**Review by JonasBaeumer — CHANGES_REQUESTED** [review]
Screenshot at Jul 06 13-23-31

When loading the initial screenshot it still shows me placeholder UI components for the PubSub system.

File: Notification/PubSub.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            return []
+
+        # 1. Papers einfügen
+        for paper in papers:
+            try:
+                paper_id = paper["id"]
+                info = paper["info"]
+                embedding = embed_string(info)
+
+                db.collection.upsert(
```
Comment (JonasBaeumer) [inline Notification/PubSub.py:54]: You are not storing all relevant information from OA in the database which could lead to data inconsistencies. For example metadata such as fwci, citations, etc.

File: Notification/PubSub.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                )
+            except Exception as e:
+                logger.error(f"Fehler beim Einfügen von Paper: {e}")
+
+        # 2. Nutzerbeschreibung aus dem ersten Element ziehen
+        user_description = user_query
+        user_embedding = embed_string(user_description)
+
+        # 3. Ähnlichkeitssuche
+        similar_ids = db.perform_similarity_search(k=k, user_profile_embedding=user_embedding)
```
Comment (JonasBaeumer) [inline Notification/PubSub.py:68]: Because here you are using our defined interface method (good!)

File: Notification/PubSub.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            paper["similarity_rank"] = rank
+            results.append(paper)
+    return results
+
+
+def split_ranked_papers(
+    result: list[dict],
+    max_top_priority: int = MAX_NUMBER_OF_SENT_MAIL
+) -> tuple[list[dict], list[dict]]:
+    """
```
Comment (JonasBaeumer) [inline Notification/PubSub.py:107]: Please make the method documentation in english

File: Notification/PubSub.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    top_priority = []
+    rest = []
+
+    for paper in result:
+        rank = paper.get("similarity_rank")
+        if rank is None or rank >= 100:
+            continue  # ignorieren
+
+        if rank < 5 and len(top_priority) < max_top_priority:
```
Comment (JonasBaeumer) [inline Notification/PubSub.py:127]: Your method description says top_priority should contain the top 10 papers, however here you restrict the rank to top 5, which one is correct?

File: Notification/PubSub.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+
+def split_ranked_papers(
+    result: list[dict],
+    max_top_priority: int = MAX_NUMBER_OF_SENT_MAIL
+) -> tuple[list[dict], list[dict]]:
+    """
+    Teilt die ranked result Liste in zwei Gruppen:
+    - Liste A: Top-Papers mit Rank < 10, maximal max_top_priority Stück
+    - Liste B: Restliche Papers mit Rank < 100
```
Comment (JonasBaeumer) [inline Notification/PubSub.py:110]: How is this limited to <100? There seems to be no restriction on this method since you just append everything. I think this approach would also make more sense because I cant see why a limit of 100 papers would have any additional benefit.

File: Notification/PubSub.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+from llm.Embeddings import embed_string, embed_user_profile
+from typing import List, Dict, Optional
+import sys
+import os
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
+
+
+logger = logging.getLogger(__name__)
+
+USER_QUERY = "I'm researching the application of Gaussian Process Bandits in optimizing machine learning tasks. Specifically, I'm interested in enhancing exploration-exploitation trade-offs and improving computational efficiency through GPU acceleration for large-scale experiments."
```
Comment (JonasBaeumer) [inline Notification/PubSub.py:13]: Please declare these things `USER_QUERY` and `TRIGGER_DATE` and `TAGS` and `MAX_NUMBER_OF_SEND_MAIL` to the main method since this is only used for testing. Otherwise this might confuse someone who looks at the file.

File: Notification/PubSub.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        if rank < 5 and len(top_priority) < max_top_priority:
+            top_priority.append(paper)
+        else:
+            rest.append(paper)
+
+    return top_priority, rest
+
+
+if __name__ == '__main__':
+    # useless because we have it already in our pipline. it's just a hardcode
```
Comment (JonasBaeumer) [inline Notification/PubSub.py:136]: what do these comments mean? If we use it for testing this file and its method it is not "useless" please declare it accordingly

**Comment by JonasBaeumer** [comment]
Screenshot at Jul 06 21-28-24

When I refresh my page to see the first pub/sub notifications, do we want that the same papers are shown in the recommended and the pub/sub section? This would kind of make the idea of pub/sub irrelevant. We should check that papers who are displayed in recommended do not appear in pubsub and vice versa

File: database/projectpaper_database_handler.py
Comment (JonasBaeumer) [inline database/projectpaper_database_handler.py:1]: Please add method descriptions for all these methods

File: database/projects_database_handler.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    conn.close()
+    return project_id
+
+
+def add_queries_to_project_db(queries: list[str], project_id: str):
+    conn = connect_to_db()
+    cursor = conn.cursor()
+    queries_str = repr(queries)
+
+    print("Updating projects_table with new queries")
```
Comment (JonasBaeumer) [inline database/projects_database_handler.py:35]: we should use a logger here to ensure the information is persistent

File: llm/Prompts.py [hunk 9]
```diff
@@ -106,39 +106,20 @@
...
     - First, analyze the user input for clarity, scope, and quality.
     - If it's invalid or irrelevant, use detect_out_of_scope_query and return an empty JSON.
     - If it’s vague, use reformulate_query.
     - If it’s too narrow or no good results were found previously, use retry_broaden.
     - If it’s already suitable, use accept.
-    - Once a valid and optimized query is available, always run update_papers, then get_best_papers.
+    - Once a valid and optimized query is available, always run update_papers, then get_best_papers and finally store_papers_for_project.
     - Do not fabricate paper content. Only use output from get_best_papers.
 
     💬 Output Format:
```
Comment (JonasBaeumer) [inline llm/Prompts.py:121]: Why did you remove the output format description? Isn't this a guardrail we put in place to protect us against malformatted responses from the Agent?

File: llm/tools/paper_handling_tools.py [hunk 10]
```diff
@@ -7,22 +7,54 @@
...
+    The papers are obtained by the result of get_best_papers. Use their abstract and
+    the original user prompt to generate a summary explaining why they are relevant to the
+    user.
+    Args:
+        project_id: The project the papers will be linked to.
+        papers: A list of dicts containing the papers' hashes and a project specific summary.
+            Each dict contains the following fields:
+                paper_hash: The hash of the paper. Provided by get_best_papers
+                summary: A summary describing why this paper is relevant for the user.
+                Must explain the key details of why this paper is relevant to the user without being overly verbose.
```
Comment (JonasBaeumer) [inline llm/tools/paper_handling_tools.py:32]: Remove unnecessary empty rows

File: llm/tools/paper_handling_tools.py [hunk 11]
```diff
@@ -52,6 +86,11 @@ def update_papers(queries: list[str]) -> str:
 
         status_postgres, deduplicated_papers = insert_papers(fetched_papers)
         # todo print how many new papers for debugging
+        print("Adding queries for project", project_id)
```
Comment (JonasBaeumer) [inline llm/tools/paper_handling_tools.py:89]: use logger here please

File: llm/tools/paper_handling_tools.py [hunk 12]
```diff
@@ -83,8 +122,79 @@ def update_papers(queries: list[str]) -> str:
         return ("Paper database has been updated with the latest papers & embeddings. There were some errors. "
```
Comment (JonasBaeumer) [inline llm/tools/paper_handling_tools.py:122]: Why are you not dealing with error in any way? Not all error will be caught this way, this could still lead to a crash of the system if you dont include error handling

File: llm/Prompts.py (hunk 9 above)
Comment (AndresDChB) [reply llm/Prompts.py:121]: Right now the agent doesnt output anything to the user. It just updates databases and the actual JSON is formed when querying the databases

File: llm/tools/paper_handling_tools.py (hunk 12 above)
Comment (AndresDChB) [reply llm/tools/paper_handling_tools.py:122]: This has been the behaviour of even the old tools. We need to add error handling in another PR
